frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -636,7 +726,65 @@ function DataFrame({
               // This results in the first cell being selected for a short period of time
               // But for touch devices, preventing this can cause issues to select cells.
               // So we allow selection changes for touch devices even when it is not focused.
-              setGridSelection(newSelection)
+              const rowSelectionChanged = !isEqual(
+                newSelection.rows.toArray(),
+                gridSelection.rows.toArray()
+              )
+
+              const columnSelectionChanged = !isEqual(
+                newSelection.columns.toArray(),
+                gridSelection.columns.toArray()
+              )
+
+              let updatedSelection = newSelection
+              if (
+                (isRowSelectionActivated || isColumnSelectionActivated) &&
+                newSelection.current !== undefined
+              ) {
+                // The default behavior is that row selections are cleared when a cell is selected.
+                // This is not desired when row selection is activated. Instead, we want to keep the
+                // row selection and only update the cell selection.
+                updatedSelection = {
+                  ...newSelection,
+                  rows: gridSelection.rows,
+                  columns: gridSelection.columns,
+                }
+              }
+
+              if (
+                rowSelectionChanged &&
+                newSelection.rows.length > 0 &&
+                columnSelectionChanged &&
+                newSelection.columns.length === 0
+              ) {
+                // Keep the column selection if row selection was changed
+                updatedSelection = {
+                  ...updatedSelection,
+                  columns: gridSelection.columns,
+                }
+              }
+              if (
+                columnSelectionChanged &&
+                newSelection.columns.length > 0 &&
+                rowSelectionChanged &&
+                newSelection.rows.length === 0
+              ) {
+                // Keep the row selection if column selection was changed
+                updatedSelection = {
+                  ...updatedSelection,
+                  rows: gridSelection.rows,
+                }
+              }
+
+              setGridSelection(updatedSelection)
+
+              if (
+                (isRowSelectionActivated && rowSelectionChanged) ||
+                (isColumnSelectionActivated && columnSelectionChanged)
+              ) {
+                applySelections(updatedSelection)
```

Comment:
nit: to make it clearer what this function is doing vs `setGridSelection`, we could call it something like `syncWithBackend` or something like that

Reply:
`syncWithBackend` might also a bit misleading since there are a couple of situations where it doesn't sync with backend. E.g. in form or if nothing has changed. I renamed it to `synceSelectionState`